### PR TITLE
executor: support the remained types for max/min

### PR DESCRIPTION
**Pull request by XuHuaiyu**
## What have you changed? (mandatory)

support String/ Time/ Duration/ JSON for max/min

## What is the type of the changes? (mandatory)

- Improvement

## How has this PR been tested? (mandatory)

The existing test cases.
randgen-test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Does this PR need to be added to the release notes? (mandatory)

no

## Refer to a related PR or issue link (optional)
#6952

### CODE REVIEW

**Comment by zz-jason** [comment]
LGTM

**Comment by crazycs520** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
I tried these ways @zz-jason 
```
func BenchmarkCopyString(b *testing.B) {
	a := "123"
	for i := 0; i < b.N; i ++{
		a = fmt.Sprintf("%s", a)
	}
}

func BenchmarkCopyString2(b *testing.B) {
	a := "123"
	for i := 0; i < b.N; i ++{
		a = a + ""
	}
}

func BenchmarkCopyString3(b *testing.B) {
	a := "123"
	for i := 0; i < b.N; i ++{
		bytes := []byte(a)
		b := make([]byte, len(bytes))
		copy(b,bytes )
		a = string(b)
	}
}

func BenchmarkCopyString4(b *testing.B) {
	a := "123"
	for i := 0; i < b.N; i ++{
		pbytes := (*reflect.SliceHeader)(unsafe.Pointer(&b))
		b := make([]byte, pbytes.Len)
		copy(b, a)
		a = string(b)
	}
}
```
BenchmarkCopyString-4           10000000               132 ns/op              19 B/op          2 allocs/op
BenchmarkCopyString2-4          100000000               11.4 ns/op             0 B/op          0 allocs/op
BenchmarkCopyString3-4          30000000                45.6 ns/op             6 B/op          2 allocs/op
BenchmarkCopyString4-4               200          10791498 ns/op        52690966 B/op          2 allocs/op

**Comment by zz-jason** [comment]
@XuHuaiyu `BenchmarkCopyString4` is not correct

**Comment by crazycs520** [comment]
@XuHuaiyu BenchmarkCopyString2 is not correct, it have not copy

```go
import . "github.com/pingcap/tidb/util/hack"
func TestStringcs(t *testing.T) {
	a := "abc"
	b := a + ""
	bufA := Slice(a)
	bufB := Slice(b)
	fmt.Println(&bufA[0], &bufB[0]) // the adress is same
}

File: executor/aggfuncs/func_max_min.go [hunk 1]
```diff
@@ -290,9 +311,202 @@ func (e *maxMin4Decimal) UpdatePartialResult(sctx sessionctx.Context, rowsInGrou
 			continue
 		}
 		cmp := input.Compare(&p.val)
-		if e.isMax && cmp == 1 || !e.isMax && cmp == -1 {
+		if e.isMax && cmp > 0 || !e.isMax && cmp < 0 {
```
Comment (crazycs520) [inline executor/aggfuncs/func_max_min.go]: Why only change this line? line#363,411?

Comment (XuHuaiyu) [reply executor/aggfuncs/func_max_min.go]: caused by the implementation of different `compare` function.

File: util/stringutil/string_util.go [hunk 2]
```diff
@@ -234,3 +235,8 @@ func DoMatch(str string, patChars, patTypes []byte) bool {
 	}
 	return sIdx == len(str)
 }
+
+// CopyString deep copies a string.
+func CopyString(src string) string {
```
Comment (zz-jason) [inline util/stringutil/string_util.go]: How about `s/CopyString/Copy/`? We usually call the function with the package name, "stringutil.Copy()" would be simpler.

File: util/stringutil/string_util.go [hunk 3]
```diff
@@ -234,3 +235,8 @@ func DoMatch(str string, patChars, patTypes []byte) bool {
 	}
 	return sIdx == len(str)
 }
+
+// Copy deep copies a string.
```
Comment (crazycs520) [inline util/stringutil/string_util.go]: add comment why we need to copy a string. because the string is `hack` from byte(chunk)

Comment (zz-jason) [reply util/stringutil/string_util.go]: no need to add it here, should add the comments in the caller side.

**Pull request merged**